ipykernel/subshell_manager.py

```diff
@@ -0,0 +1,280 @@
+        return self._control_other_socket
+
+    def get_other_socket(self, subshell_id: str | None) -> zmq.asyncio.Socket:
+        """Return the other inproc pair socket for a subshell.
+
+        This socket is accessed from the subshell thread.
+        """
+        if subshell_id is None:
+            return self._parent_other_socket
+        with self._lock_cache:
+            socket = self._cache[subshell_id].thread._pair_socket
+            assert socket is not None
+            return socket
+
+    def get_shell_channel_socket(self, subshell_id: str | None) -> zmq.asyncio.Socket:
+        """Return the shell channel inproc pair socket for a subshell.
+
+        This socket is accessed from the shell channel thread.
+        """
+        if subshell_id is None:
+            return self._parent_shell_channel_socket
+        with self._lock_cache:
+            return self._cache[subshell_id].shell_channel_socket
+
+    def list_subshell(self) -> list[str]:
+        """Return list of current subshell ids.
+
+        Can be called by any subshell using %subshell magic.
+        """
+        with self._lock_cache:
+            return list(self._cache)
+
+    async def listen_from_control(self, subshell_task) -> None:
+        """Listen for messages on the control inproc socket, handle those messages and
+        return replies on the same socket.  Runs in the shell channel thread.
+        """
+        assert current_thread().name == "Shell channel"
+
+        socket = self._control_shell_channel_socket
+        while True:
+            request = await socket.recv_json()  # type: ignore[misc]
+            reply = await self._process_control_request(request, subshell_task)
+            await socket.send_json(reply)  # type: ignore[func-returns-value]
+
+    async def listen_from_subshells(self) -> None:
+        """Listen for reply messages on inproc sockets of all subshells and resend
+        those messages to the client via the shell_socket.
+
+        Runs in the shell channel thread.
+        """
+        assert current_thread().name == "Shell channel"
+
+        async with create_task_group() as tg:
+            tg.start_soon(self._listen_for_subshell_reply, None)
+            async for subshell_id in self._receive_stream:
+                tg.start_soon(self._listen_for_subshell_reply, subshell_id)
+
+    def subshell_id_from_thread_id(self, thread_id) -> str | None:
+        """Return subshell_id of the specified thread_id.
+
+        Raises RuntimeError if thread_id is not the main shell or a subshell.
+
+        Only used by %subshell magic so does not have to be fast/cached.
+        """
+        with self._lock_cache:
+            if thread_id == main_thread().ident:
+                return None
+            for id, subshell in self._cache.items():
+                if subshell.thread.ident == thread_id:
+                    return id
+            msg = f"Thread id '{thread_id} does not correspond to a subshell of this kernel"
```

Comment:
Unmatched ', trying to ass quotes ?  Maybe `!r` ?
```suggestion
            msg = f"Thread id {thread_id!r} does not correspond to a subshell of this kernel"
```

Reply:
This was probably just a typo on my behalf!

Reply:
Fixed ind17c77f.